packages/lit-dev-content/src/components/litdev-search.ts

```diff
@@ -81,13 +81,73 @@ interface SuggestionGroup {
 }
 
 /**
- * A collection of SuggestionGroups in which the key is determined by the
- * suggestion title and the type of content.
- *
- * e.g. LifecycleDocs or ReactiveElementAPI
+ * A data structure that takes in a list of Suggestions, groups them by
+ * parentID, reformats the data for rendering, and then de-ranks the tutorial
+ * results.
  */
-interface SuggestionGroups {
-  [group: string]: SuggestionGroup;
+class SuggestionGroups {
+  private orderedPageIds: string[] = [];
+  private orderedTutorialPageIds: string[] = [];
+  private pageIdToSuggestionGroup: Map<string, SuggestionGroup> = new Map();
+
+  constructor(suggestions: Suggestion[]) {
+    suggestions.forEach((suggestion) => this.add(suggestion));
+  }
+
+  /**
+   * Adds a suggestion to the data structure. If the suggestion is a tutorial,
+   * it is added to the end of the list of tutorials. Otherwise, it is added to
+   * the end of it's corresponding group. If there is no corresponding group,
+   * one will be created. (performance O(suggestions))
+   *
+   * @param suggestion Suggestion to add to data structure.
+   */
+  add(suggestion: Suggestion) {
+    const id = suggestion.parentID ?? suggestion.objectID;
+    const suggestionGroup = this.pageIdToSuggestionGroup.get(id) ?? {
+      ...suggestion.docType,
+      suggestions: [],
+    };
+
+    // This is the first time we've encountered this parentID
+    if (!suggestionGroup.suggestions.length) {
+      this.pageIdToSuggestionGroup.set(id, suggestionGroup);
+
+      // If this is a tutorial, add it to the end of the tutorial list which
+      // will be rendered after the other suggestions.
+      if (suggestion.docType.tag === 'tutorial') {
+        this.orderedTutorialPageIds.push(id);
+      } else {
+        this.orderedPageIds.push(id);
+      }
+    }
+
+    if (!suggestion.parentID) {
+      // If it's the title link, prepend it so it shows up first.
+      suggestionGroup.suggestions.unshift(suggestion);
+    } else {
+      suggestionGroup.suggestions.push(suggestion);
+    }
+  }
+
+  /**
+   * Gets a suggestion group by it's parentID.
+   *
+   * @param key The parentID of the suggestion group.
+   * @returns The suggestion group associated with the given parentID.
+   */
+  get(key: string) {
+    return this.pageIdToSuggestionGroup.get(key);
+  }
+
+  /**
+   * Get all the keys, or parentIDs, of the suggestion groups.
+   *
+   * @returns An array of ordered parentIDs where tutorial IDs are at the end.
+   */
+  keys(): string[] {
```

Comment:
Nit: `keys()`is used to pass an iterable of `groupID`'s into `repeat`. Every time it's allocating a new list. The class `SuggestionGroups` could implement [`Symbol.iterator`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/iterator) and be passed directly to the first argument of `repeat`. Maybe something like: [this playground](https://lit.dev/playground/#project=W3sibmFtZSI6InNpbXBsZS1ncmVldGluZy50cyIsImNvbnRlbnQiOiJjb25zdCBhcnIxID0gWycxJywgJzEwJ107XG5jb25zdCBhcnIyID0gWydoZXkhJ107XG5cbmNvbnN0IGl0ZXJhYmxlMSA9IHtcbiAgW1N5bWJvbC5pdGVyYXRvcl06IGZ1bmN0aW9uKiAoKSB7XG4gICAgeWllbGQqIGFycjFbU3ltYm9sLml0ZXJhdG9yXSgpO1xuICAgIHlpZWxkKiBhcnIyW1N5bWJvbC5pdGVyYXRvcl0oKTtcbiAgfVxufTtcblxuZm9yIChjb25zdCBpdGVtIG9mIGl0ZXJhYmxlMSkge1xuICBjb25zb2xlLmxvZyhpdGVtKTtcbn1cbiJ9LHsibmFtZSI6ImluZGV4Lmh0bWwiLCJjb250ZW50IjoiPCFET0NUWVBFIGh0bWw-XG48aGVhZD5cbiAgPHNjcmlwdCB0eXBlPVwibW9kdWxlXCIgc3JjPVwiLi9zaW1wbGUtZ3JlZXRpbmcuanNcIj48L3NjcmlwdD5cbjwvaGVhZD5cbjxib2R5PlxuICA8c2ltcGxlLWdyZWV0aW5nIG5hbWU9XCJXb3JsZFwiPjwvc2ltcGxlLWdyZWV0aW5nPlxuPC9ib2R5PlxuIn0seyJuYW1lIjoicGFja2FnZS5qc29uIiwiY29udGVudCI6IntcbiAgXCJkZXBlbmRlbmNpZXNcIjoge1xuICAgIFwibGl0XCI6IFwiXjIuMC4wXCIsXG4gICAgXCJAbGl0L3JlYWN0aXZlLWVsZW1lbnRcIjogXCJeMS4wLjBcIixcbiAgICBcImxpdC1lbGVtZW50XCI6IFwiXjMuMC4wXCIsXG4gICAgXCJsaXQtaHRtbFwiOiBcIl4yLjAuMFwiXG4gIH1cbn0iLCJoaWRkZW4iOnRydWV9XQ). That would then iterate over both array iterators without allocating an intermediate list (I think?).

Reply:
Ah I prototyped returning the actual suggestions in order but found that was actually too gross and complex. Your approach is much nicer